graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -169,3 +127,148 @@ def merge_schemas(schemas_dict):
+    return blank_ast
+
+
+def _process_directive_definition(directive, existing_directives, merged_schema_ast):
+    """Compare new directive against existing ones, update records and schema accordingly.
+
+    If the directive is new, in that no existing directive shares its name, it will be added
+    to both existing_directives and merged_schema_ast.
+    If the directive already exists, in that there is some existing directive that shares its
+    name and structure, nothing will be changed.
+    If there is a conflict, in that there is some existing directive that shares its name but
+    has a different structure, an error will be raised.
+
+    existing_directives and merged_schema_ast may be modified.
+
+    Args:
+        directive: DirectiveDefinition, an AST node representing the definition of a directive
+        existing_directives: Dict[str, DirectiveDefinition], mapping the name of each existing
+                             directive to the AST node defining it
+        merged_schema_ast: Document, AST representing a schema
+
+    Raises:
+        - SchemaNameConflictError if there is an existing directive with the same name, but
+          a different definition, as the input directive
+    """
+    directive_name = directive.name.value
+    if directive_name in existing_directives:
+        if directive == existing_directives[directive_name]:  # definitions agree
+            return
+        else:  # definitions disagree
+            raise SchemaNameConflictError(
+                u'Directive "{}" with definition "{}" has already been defined with '
+                u'definition "{}".'.format(
+                    directive_name,
+                    print_ast(directive),
+                    print_ast(existing_directives[directive_name]),
+                )
+            )
+    # new directive
+    merged_schema_ast.definitions.append(directive)  # Add to AST
+    existing_directives[directive_name] = directive  # Add to record of directives
+
+
+def _process_scalar_definition(scalar, existing_scalars, existing_types, merged_schema_ast):
+    """Compare new scalar against existing scalars and types, update records and schema.
+
+    If the scalar is new and does not conflict with any existing type, it will be added to both
+    existing_scalars and merged_schema_ast.
+    if the scalar is new but clashes with the name of some existing type, an error will be raised.
+    If the scalar already exists, nothing will be changed.
+
+    existing_scalars and merged_schema_ast may be modified.
+
+    Args:
+        scalar: ScalarDefinition, an AST node representing the definition of a scalar
+        existing_scalars: Set[str], set of names of all existing scalars
+        existing_types: Set[str], set of names of all existing types
+        merged_schema_ast: Document, AST representing a schema
+
+    Raises:
+        - SchemaNameConflictError if there is an existing type those name conflicts with the
+          new scalar
+    """
```

Comment:
I am against this much documentation for this function. It is easier for me to read the code than the documentation. I would leave the documentation as a one section headline.

Reply:
I actually found it easier to understand with the fleshed-out docstring, although I think it's valid to take out the paragraph right below the headline since the code is pretty readable below.

I think keeping the argument types (Args) and Raises section would be nice, and we can take out `# Add to AST` and `# Add to record of scalars` down below.

Reply:
I also rather like documenting the types of the arguments, because it's not super trivial to find out. I'm putting the args back for now, deleting some useless comments elsewhere.

Reply:
Ok. This decision makes sense to me.